### ISSUE TITLE
PrettyWriter shouldn't publicly inherit from Writer and then hide its methods

### ISSUE THREAD

**Issue by tonyelewis** [miloyip/rapidjson#912]
From what I can see, `PrettyWriter` publicly inherits from `Writer` and hides some of its methods (eg `Int()`). I believe this sort of thing is generally thought to be a bad idea. In particular, the `public` inheritance advertises to users that `PrettyWritter` can be safely/correctly accessed through a `Writer` pointer or reference but that's untrue&mdash;accessing a `PrettyWriter` through a `Writer` pointer/reference leads to the `Writer` versions of the relevant methods being called. This could lead to all sorts of inconsistency and/or confusion.

I presume the methods are hiding rather overriding to avoid the cost of a `virtual` lookup, which seems reasonable. But then:
 * I think this would be better modeled using `protected` or `private` inheritance.
 * If the purpose of the inheritance is to share common functionality, it may be better to abstract that code into a separate class that's used by both `Writer` and `PrettyWriter`.
 * Either way, it may be better to consider avoiding inheritance altogether, making `PrettyWriter` share another class's code though a member of that other class instead (ie ["Prefer Composition over Inheritance"](https://en.wikipedia.org/wiki/Composition_over_inheritance)).

I'm not terribly familiar with the rapidjson code; I apologise if these comments aren't helpful.

**Comment by AndyNicholasShield** [comment]
I totally agree. I'm trying to pass a single write object to routines which do the writing of the json via the generic Writer<> class. I assumed I could create a PrettyWriter<> class and a Writer<> class and choose between the "writer" to send to the methods which did the writing... but no! The code which does the writing never does the pretty-printing.

I assume the only way to avoid writing this code twice is... write it twice? Once for pretty and once for regular non-pretty?

If something needs to change, I would make both writers completely separate classes.
